### ISSUE TITLE
Add `FlxG.warning()` or `FlxG.error()`

### ISSUE THREAD

**Issue by IQAndreas** [FlixelCommunity/flixel#89]
Currently Flixel does the following when "reporting" an error:

``` as3
FlxG.log("WARNING: Paths need at least one node in them to be followed.");
```

It would be clearer being able to do one of these (and avoids the repetitious "WARNING:" prefix):

``` as3
FlxG.warn(this, "Paths need at least one node in them to be followed.");
FlxG.error(this, "Cannot add a `null` object to a FlxGroup.");
```

This allows you to handle errors differently if you are debugging or if the error occurs when it's running on a server. Or perhaps you want all errors logged to a server rather than the Flash console.

(passing `this` allows you to do several neat things, including showing where the error occurred and simple variable dumps)

**Comment by Dovyski** [comment]
I like the idea. This can produce "red" or "yellow" text lines in Flixel console, something much easier to notice as a warning/error.

**Comment by Gama11** [comment]
Might also simply have a second, optional parameter in FlxG.log specifying the color along with a warn()-function.

**Comment by Dovyski** [comment]
It's also a good idea :)

For the sake of simplicity isn't easier to add a second parameter to `FlxG.log` that tells the "type" of log?

``` as3
FlxG.log("warning stuff", WARNING);
FlxG.log("errorstuff", ERROR);
FlxG.log("usual log"); // defaults to INFO
```

**Comment by Gama11** [comment]
> For the sake of simplicity isn't easier to add a second parameter to FlxG.log that tells the "type" of log?

Hm, yeah, that would be another option. Could have both options, since I think a `warn` and `error`-function would be good to differentiate them more clearly. Even if warn does nothing else but calling `FlxG.log` with certain paramters, for example: 

``` actionscript
public function warn(text:String):void
{
     log(text, FlxG.RED);
}
```

In case log takes a second, color-paramter. The question is, is color the only thing where warn differs from log? Could also be bold... So log could have a third paramter for font style (NORMAL, ITALIC, BOLD). Also, error and warn could do more than just calling log, playing a sound effect and / or opening the debug menu if in debug mode could be helpful so you notice the issue in first place. Cause otherwise, the "warning / error" kinda loses its purpose.

**Comment by Dovyski** [comment]
Yeah that's true. I would implement things like you described, a function to actually do the logging (e.g. `FlxG.doLog()`) and all other functions calling that one with appropriate parameters.

The only advantage I see creating `FlxG.warn()` and `FlxG.error()` is that is makes things easier to understand, because  `FlxG.log(.., TYPE)` is equally powerful and still easy to understand.

**Comment by Gama11** [comment]
Though that'd not be as powerful as something like 

``` actionscript
public function log(Text:String, TextColor:uint = FlxG.BLACK, TextStyle:int = NORMAL, PlayErrorSound:Boolean = false, OpenDebugWindow:Boolean = false):void
```

which would be great to have IMO. But yeah, my point was mostly that it's easier to understand that way.

**Comment by Dovyski** [comment]
Oh, yeah, not as powerful as that `log()` on steroids :)

**Comment by IQAndreas** [comment]
The thing I really like about `trace()` is that you can pass any number of parameters and it will concatenate them with a space.

This means you can do things like this:

``` as3
// For warnings/errors
trace(this, "Cannot set", variableName, "to the value", badValue);

// For debugging
trace("[FlxSprite::update]", indexX, indexY, widthHelper, _colorTransform);
```

Which would normally require several `+ " " +` between the values.

So, [whenever I write a log function](https://gist.github.com/4450713), I always allow any number of arguments. I would like to keep that functionality.

**Comment by IQAndreas** [comment]
Perhaps we could have an `advancedLog()` function?

``` as3
public function advancedLog(Text:String, Priority:uint = FlxLog.WARNING, TextColor:uint = FlxG.BLACK, TextStyle:int = NORMAL):void
```

And the "simpler" functions (with any number of arguments) are simply "frontends" for this advanced log function:

``` as3
public function warn(... args):void
{
    var message:String = "[WARNING] " + (args.length > 0 ? args.join(" ") : "");
    advancedLog(message, FlxLog.WARNING, FlxG.YELLOW, NORMAL);
}
```

As for the difference between warning and error, I think they should all be classified "priorities". Then we can add a system that lets you "listen" for errors, passing a callback function, and only running that function if the error is above a certain priority.

For instance, the suggestion by @Gama11 to play a sound and show the console could be done by this callback system:

``` as3
FlxLog.listen(FlxLog.ERROR, onErrorOrAbove);
function onErrorOrAbove(message:String, data:*)
{
    errorSound.play();
    FlxG.console.show(); // I'm not sure how to _actually_ show the console. I'm just guessing here.
    sendToServer(message, data);
}
```

**Comment by Gama11** [comment]
I've personally never used `trace()` in that way, but I definitely agree that the multi-parameter-solution seems nicer that the use of tons and tons of `+`.

When I just read about that suggestion, I thought it wouldn't really work very well in conjunction with all the additional paramters I suggested, but `advancedLog()` seems like a great solution.

Having the ability so set a callback for different priorities sounds like a great idea - but that should be optional. I think it's ok to show the console and play a sound effect for the highest priority (error) and maybe just a sound effect for warnings? Not sure.

So, `error` and `warn` would need if / else logic, checking whether `errorCallback` / `warnCallback` is `null` and acting accordingly, either calling the callback function or executing the default behaviour for that priority level.

Another mechanism that could be helpful is flixel automatically writing errors and warnings into a local shared object, a logfile basically, to make sure you don't miss them. You should be able to read those easily with a tool like soledit... I think `FlxSave` got some encryption going on?

What happens when `debug = false`? Are logs simply ignored? Flixel 2.5+ seems to auto-disable debugging when the swf is hosted online. 

How would settting callback functions be handled? I'd go for something like

``` actionscript
public function setLogCallback(Priority:uint, Callback:Function):void
```

**Comment by Gama11** [comment]
Basically, with all these new functions, the use of logging is encouraged a lot:

``` actionscript
FlxG.warn()
FlxG.error()
FlxG.log()
FlxG.advancedLog()
FlxG.setLogCallback()
```

However, I never really liked the Flixel 2.5+ console. While the addition of `FlxG.watch`, the `visualDebug`-button and the additional `stats` are great, the 'watch' and the `log` window are both extremely small by default. I prefered older flixel versions in that regard, where the `log` textfield basically covered the whole screen. 

The main problem I have with the console is the use of `FlxWindow` - it's nothing more than unneccessary UI that clutters things in my opinion. Why do you need to move the windows around when you can change the layout with `FlxG.setDebuggerLayout` anway? I guess it the way the windows are arranged would be saved, but if it's reset everytime you reload the .swf, it doesn't make much sense. The only time I ever used the resizing / moving feature of those windows is by accident - I think that can even happen while the console isn't opened.
Other than that, it doesn't seem to be implemented very well. When two windows are overlapping, both will be resized / moved, while you'd typically expect only the top window is affected. Also, you can't move windows closer to the edges than a few pixels.

Now, I realize others might feel different about those windows, I'd love to hear your opinions, but I'd suggest at least adding a `DEBUGGER_CLASSIC`-layout which has static textfields similar to Flixel versions lower than 2.5. The left half of the screen would be used for `log`, the right side for `watch`.

**Comment by Dovyski** [comment]
I liked the idea of `advancedLog`, but I think the log callback should be added the way @Gama11 described: something like `FlxG.setLogcallback()` or `FlxG.addLogCallback()`. It is easier to understand. The addition of `advancedLog` and callbacks will allow developers to do some cool things like integrating Flixel with [Monsterdebugger](http://demonsterdebugger.com/).

Regaring Flixel console, I never used anything before the 2.5 version. I do agree the console window is too small. Most of the time, I don't move it around (only when I want to make it larger). The watch window makes sense only if there is something being watched, in my opinion.

**Comment by Gama11** [comment]
> Regaring Flixel console, I never used anything before the 2.5 version. I do agree the console window is too small. Most of the time, I don't move it around (only when I want to make it larger). The watch window makes sense only if there is something being watched, in my opinion.

Glad you agree with me there. Only showing the watch window if `FlxG.watch()` has been called sound like a good idea to me. 
Another thing that might work is deactiving the window behaviour by default, but leaving the option to activate it via a Boolean like `FlxG.flexibleConsoleWindows` /  `FlxG.flexibleDebugWindows`.

Shall I create a new issue for these changes to the console, as it doesn't really belong here?

**Comment by Dovyski** [comment]
Yes, I think you should create another issue, otherwise we will generate too many off-topic discussions here.
